src/skdh/io/csv.py

```diff
@@ -0,0 +1,346 @@
+"""
+CSV reading process
+
+Lukas Adamowicz
+Copyright (c) 2023. Pfizer Inc. All rights reserved
+"""
+from warnings import warn
+
+from numpy import tile, arange, mean, diff, asarray, argmin, abs, vstack, unique, all as npall, int_
+from pandas import read_csv, to_datetime, to_timedelta, Timedelta
+
+from skdh.base import BaseProcess
+from skdh.io.base import check_input_file
+
+
+def handle_timestamp_inconsistency(df, fill_gaps, accel_col_names, accel_in_g, g):
+    """
+    Handle any time gaps, or timestamps that are only down to the second.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        Dataframe containing timestamps and acceleration data.
+    fill_gaps : bool
+        Fill data gaps with the vector [0, 0, 1] or [0, 0, `g`] depending on
+        accel units.
+    accel_col_names : array-like
+        Array-like (size 3) of the acceleration column names in XYZ order.
+    accel_in_g : bool
+        If acceleration values are already in units of "g".
+    g : float
+        Gravitational acceleration in m/s^2.
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        Dataframe with corrected timestamps.
+    fs : float
+        Number of samples per second.
+    """
+    # get a sampling rate. If non-unique timestamps, this will be updated
+    n_samples = mean(1 / diff(df['_datetime_'][:2500]).astype(int)) * 1e9  # datetime diff is in ns
+
+    # first check if we have non-unique timestamps
+    nonuniq_ts = df["_datetime_"].iloc[1] == df["_datetime_"].iloc[0]
+
+    # if there are non-unique timestamps, fix
+    if nonuniq_ts:
+        # check that all the blocks are the same size (or that there is only 1 non-equal block
+        # at the end)
+        _, counts = unique(df["_datetime_"], return_counts=True)
+        if not npall(counts[:-1] == counts[0]):
+            raise ValueError(
+                "Blocks of non-unique timestamps are not all equal size. "
+                "Unable to continue reading data."
+            )
+        # check if the last block is the same size
+        if counts[-1] != counts[0]:
+            # drop the last blocks worth of data
+            warn("Non integer number of blocks. Trimming partial block.", UserWarning)
+            df = df.iloc[0:-counts[-1]]
+
+        # get the number of samples, and the number of blocks
+        n_samples = counts[0]
+        n_blocks = df.shape[0] / n_samples
+
+        # compute time delta to add
+        t_delta = tile(arange(0, 1, 1 / n_samples), int(n_blocks))
+        t_delta = to_timedelta(t_delta, unit='s')
+
+        # add the time delta so that we have unique timestamps
+        df['_datetime_'] += t_delta
+
+    # now fix any data gaps: set the index as the datetime, and then upsample to match
+    # the sampling frequency. This will put nan values in any data gaps
+    df_full = df.set_index('_datetime_').asfreq(f'{1 / n_samples}S')
```

Comment:
That's a good point, I'll make that change, though we would have to fail to read if we aren't interpolating missing timestamps and there are data gaps. If we don't raise an error, then the data could be passed on to later algorithms that would produce garbage outputs.

Reply:
yeah, I agree. If there is a datagap and interpolation is set to False, then it should fail and the user can decide how they want to handle it. If interpolation is set to True, it will just interpolate with static values